File: typescript/src/opf.ts
/**
 * PPTX.dev Open Source Presentation Format (OPF) v1.0
 *
 * A JSON-based file format for describing PowerPoint presentations.
 * Agents describe narratives, content, audience, and design intent —
 * the engine handles the OOXML complexity.
 *
 * Licensed under MIT
 */

// ─── Top-Level Document ──────────────────────────────────────────────

export interface OPFDocument {
  /** Identifies the OPF schema version for validators and editors */
  $schema: "https://pptx.dev/schema/opf/v1";

  /** High-level presentation metadata */
  meta: OPFMeta;

  /** Design system: colors, fonts, spacing, brand */
  design: OPFDesign;

  /** Ordered array of slides */
  slides: OPFSlide[];

  /** Optional global speaker notes / presenter guide */
  presenterGuide?: string;

  /** Optional custom data passthrough for agent workflows */
  extensions?: Record<string, unknown>;
}

// ─── Metadata ────────────────────────────────────────────────────────

export interface OPFMeta {
  title: string;
  /**
   * Optional base filename for exports (without extension). The engine
   * silently normalizes whitespace, drops a trailing `.`, and strips one
   * trailing recognized export extension (`.pptx`, `.pdf`, `.png`, `.svg`,
   * case-insensitive) before appending the target format's extension.
   * When omitted, the engine slugifies `meta.title`.
   */
  filename?: string;
  subtitle?: string;
  author?: string;
  company?: string;
  audience?: string;
  purpose?: string;
  narrative?: OPFNarrative;
  language?: string;
  createdAt?: string;
  tags?: string[];
}

export interface OPFNarrative {
  /** Reference to a named narrative template, e.g. "problem-solution" */
  template?: string;
  /** Free-form narrative description for AI-driven generation */
  description?: string;
  /** Key messages to convey */
  keyMessages?: string[];
  /** Desired tone: "formal" | "casual" | "inspirational" | "technical" | "persuasive" */
  tone?: string;
  /** Target duration in minutes */
  durationMinutes?: number;
}

// ─── Design System ───────────────────────────────────────────────────

export interface OPFDesign {
  /** Reference a named theme from pptx.gallery, or define inline */
  theme?: string;

  /** Color palette */
  colors?: OPFColorScheme;

  /** Typography */
  fonts?: OPFFontScheme;

  /** Slide dimensions */
  dimensions?: OPFDimensions;

  /** Background defaults */
  background?: OPFBackground;

  /** Logo / brand assets */
  brand?: OPFBrand;

  /** Layout preference hints */
  layoutPreferences?: OPFLayoutPreferences;
}

export interface OPFColorScheme {
  /** Named scheme from gallery, e.g. "ocean-depth", "corporate-blue" */
  scheme?: string;
  /** Override individual color roles */
  primary?: string;
  secondary?: string;
  accent?: string;
  background?: string;
  surface?: string;
  text?: string;
  textSecondary?: string;
  /** Custom named colors for advanced use */
  custom?: Record<string, string>;
}

export interface OPFFontScheme {
  /** Named scheme from gallery, e.g. "modern-sans", "classic-serif" */
  scheme?: string;
  /** Override individual font roles */
  heading?: OPFFont;
  body?: OPFFont;
  accent?: OPFFont;
  code?: OPFFont;
}

export interface OPFFont {
  family: string;
  weight?: number;
  style?: "normal" | "italic";
  letterSpacing?: number;
}

export interface OPFDimensions {
  /** Preset: "16:9" | "4:3" | "16:10" | "letter" | "a4" */
  preset?: string;
  /** Custom width in inches */
  widthInches?: number;
  /** Custom height in inches */
  heightInches?: number;
}

export interface OPFBackground {
  type: "solid" | "gradient" | "image" | "pattern";
  color?: string;
  gradient?: { angle: number; stops: { color: string; position: number }[] };
  image?: { url: string; fit: "cover" | "contain" | "tile" };
  opacity?: number;
}

export interface OPFBrand {
  logo?: { url: string; position?: OPFPosition; widthInches?: number };
  watermark?: { url: string; opacity?: number };
  companyName?: string;
}

export interface OPFLayoutPreferences {
  /** Default content density: "minimal" | "balanced" | "dense" */
  density?: string;
  /** Default alignment: "left" | "center" | "right" */
  alignment?: string;
  /** Use animations */
  animations?: boolean;
  /** Slide numbering */
  slideNumbers?: boolean;
}

// ─── Slides ──────────────────────────────────────────────────────────

export interface OPFSlide {
  /** Unique ID within the document */
  id: string;

  /** Layout type — references pptx.gallery layouts or custom */
  layout: string;

  /** Slide-level overrides to design */
  designOverrides?: Partial<OPFDesign>;

  /** Content elements on this slide */
  elements: OPFElement[];

  /** Speaker notes for this slide */
  notes?: string;

  /** Transition to this slide */
  transition?: OPFTransition;

  /** Whether this slide is hidden */
  hidden?: boolean;

  /** Section marker — groups slides in presenter view */
  section?: string;
}

export interface OPFTransition {
  type: "none" | "fade" | "slide" | "push" | "wipe" | "morph" | "zoom";
  duration?: number;
  direction?: "left" | "right" | "up" | "down";
}

// ─── Elements ────────────────────────────────────────────────────────

export type OPFElement =
  | OPFTextElement
  | OPFImageElement
  | OPFShapeElement
  | OPFChartElement
  | OPFTableElement
  | OPFVideoElement
  | OPFCodeElement
  | OPFGroupElement
  | OPFPlaceholderElement;

export interface OPFElementBase {
  id: string;
  type: string;
  position?: OPFPosition;
  size?: OPFSize;
  /** Animation on this element */
  animation?: OPFAnimation;
  /** Named slot from layout template, e.g. "title", "body", "image-right" */
  slot?: string;
}

export interface OPFPosition {
  x?: number;
  y?: number;
  /** Anchor to layout grid: "top-left" | "center" | "bottom-right" etc. */
  anchor?: string;
}

export interface OPFSize {
  width?: number;
  height?: number;
  /** Unit: "in" (inches), "%" (percentage of slide), "px" */
  unit?: "in" | "%" | "px";
}

export interface OPFAnimation {
  type: "appear" | "fadeIn" | "slideIn" | "zoomIn" | "typewriter" | "custom";
  delay?: number;
  duration?: number;
  order?: number;
}

// ─── Text ────────────────────────────────────────────────────────────

export interface OPFTextElement extends OPFElementBase {
  type: "text";
  content: OPFTextContent;
  style?: OPFTextStyle;
}

export interface OPFTextContent {
  /** Simple text string — for basic use */
  text?: string;
  /** Rich text with formatting runs */
  runs?: OPFTextRun[];
  /** Markdown source — converted to runs by engine */
  markdown?: string;
  /** Bullet list */
  bullets?: (string | OPFTextContent)[];
  /** Numbered list */
  numbered?: (string | OPFTextContent)[];
}

export interface OPFTextRun {
  text: string;
  bold?: boolean;
  italic?: boolean;
  underline?: boolean;
  strikethrough?: boolean;
  color?: string;
  fontSize?: number;
  fontFamily?: string;
  link?: string;
  superscript?: boolean;
  subscript?: boolean;
}

export interface OPFTextStyle {
  fontSize?: number;
  fontFamily?: string;
  color?: string;
  alignment?: "left" | "center" | "right" | "justify";
  lineHeight?: number;
  letterSpacing?: number;
  textTransform?: "none" | "uppercase" | "lowercase" | "capitalize";
  verticalAlignment?: "top" | "middle" | "bottom";
}

// ─── Image ───────────────────────────────────────────────────────────

export interface OPFImageElement extends OPFElementBase {
  type: "image";
  src: string;
  alt?: string;
  fit?: "cover" | "contain" | "fill" | "none";
  borderRadius?: number;
  shadow?: boolean;
  caption?: string;
}

// ─── Shape ───────────────────────────────────────────────────────────

export interface OPFShapeElement extends OPFElementBase {
  type: "shape";
  shape: "rectangle" | "circle" | "ellipse" | "triangle" | "arrow" | "line" | "star" | "callout" | "custom";
  fill?: string;
  stroke?: { color: string; width: number };
  cornerRadius?: number;
  text?: OPFTextContent;
  /** SVG path for custom shapes */
  path?: string;
}

// ─── Chart ───────────────────────────────────────────────────────────

export interface OPFChartElement extends OPFElementBase {
  type: "chart";
  chartType: "bar" | "column" | "line" | "pie" | "donut" | "area" | "scatter" | "radar" | "waterfall" | "funnel" | "treemap" | "combo";
  data: OPFChartData;
  options?: OPFChartOptions;
}

export interface OPFChartData {
  labels?: string[];
  datasets: OPFChartDataset[];
}

export interface OPFChartDataset {
  label: string;
  values: number[];
  color?: string;
  /** For combo charts */
  type?: string;
}

export interface OPFChartOptions {
  title?: string;
  legend?: boolean | { position: "top" | "bottom" | "left" | "right" };
  showValues?: boolean;
  showGrid?: boolean;
  stacked?: boolean;
  /** Y-axis config */
  yAxis?: { label?: string; min?: number; max?: number; format?: string };
  /** X-axis config */
  xAxis?: { label?: string; format?: string };
}

// ─── Table ───────────────────────────────────────────────────────────

export interface OPFTableElement extends OPFElementBase {
  type: "table";
  headers?: string[];
  rows: (string | number | OPFTextContent)[][];
  style?: OPFTableStyle;
}

export interface OPFTableStyle {
  headerBackground?: string;
  headerColor?: string;
  stripedRows?: boolean;
  borderColor?: string;
  compact?: boolean;
}

// ─── Video ───────────────────────────────────────────────────────────

export interface OPFVideoElement extends OPFElementBase {
  type: "video";
  src: string;
  poster?: string;
  autoplay?: boolean;
}

// ─── Code ────────────────────────────────────────────────────────────

export interface OPFCodeElement extends OPFElementBase {
  type: "code";
  code: string;
  language?: string;
  theme?: "dark" | "light";
  showLineNumbers?: boolean;
}

// ─── Group ───────────────────────────────────────────────────────────

export interface OPFGroupElement extends OPFElementBase {
  type: "group";
  children: OPFElement[];
}

// ─── Placeholder ─────────────────────────────────────────────────────

export interface OPFPlaceholderElement extends OPFElementBase {
  type: "placeholder";
  /** AI instruction for what to generate */
  prompt: string;
  /** Expected content type */
  expectedType?: "text" | "image" | "chart" | "table";
}

File: python/src/pptx_dev/opf.py
"""Open Presentation Format (OPF) v1.0 Pydantic models.

A JSON-based file format for describing PowerPoint presentations. Agents
describe narratives, content, audience, and design intent — the engine
handles the OOXML complexity.

The models here are the Python port of the canonical TypeScript types in
``@pptx/sdk/opf`` and the JSON schema at ``https://pptx.dev/schema/opf/v1``.
They use ``populate_by_name`` so both camelCase JSON (the wire format) and
snake_case Python attributes work on input.
"""

from __future__ import annotations

from typing import Annotated, Any, Literal, Union

from pydantic import BaseModel, ConfigDict, Field


class _OPFBase(BaseModel):
    """Shared configuration for all OPF models."""

    model_config = ConfigDict(
        populate_by_name=True,
        extra="allow",
    )


# ─── Narrative / Meta ────────────────────────────────────────────────


class OPFNarrativeBeat(_OPFBase):
    """A single narrative beat — a labeled segment of the story arc.

    Slides reference beats via ``OPFSlide.beat``.
    """

    id: str
    name: str
    description: str | None = None
    slide_count: int | None = Field(default=None, alias="slideCount")
    """Optional explicit slide count for this beat. Defaults to 1 when
    omitted; values >1 are reserved for beats that intentionally span
    multiple slides. Prefer decomposing a heavy beat into multiple beats
    over setting a high ``slideCount``."""
    layout_hint: str | None = Field(default=None, alias="layoutHint")


class OPFNarrative(_OPFBase):
    """Structured storyline metadata used by AI to shape generated content.

    Narrative declares the deck's intended story arc; slides may opt into
    beats via ``OPFSlide.beat``. The narrative does not constrain slide
    structure — validators warn on drift but never error. Slides are the
    source of truth; narrative is intent that travels with the deck.
    """

    template: str | None = None
    description: str | None = None
    key_messages: list[str] | None = Field(default=None, alias="keyMessages")
    tone: str | None = None
    duration_minutes: float | None = Field(default=None, alias="durationMinutes")
    beats: list[OPFNarrativeBeat] | None = None


class OPFMeta(_OPFBase):
    title: str
    description: str | None = None
    filename: str | None = None
    subtitle: str | None = None
    author: str | None = None
    company: str | None = None
    audience: str | None = None
    purpose: str | None = None
    narrative: OPFNarrative | None = None
    language: str | None = None
    created_at: str | None = Field(default=None, alias="createdAt")
    tags: list[str] | None = None


# ─── Design System ───────────────────────────────────────────────────


class OPFColorScheme(_OPFBase):
    scheme: str | None = None
    primary: str | None = None
    secondary: str | None = None
    accent: str | None = None
    background: str | None = None
    surface: str | None = None
    text: str | None = None
    text_secondary: str | None = Field(default=None, alias="textSecondary")
    custom: dict[str, str] | None = None


class OPFFont(_OPFBase):
    family: str
    weight: int | None = None
    style: Literal["normal", "italic"] | None = None
    letter_spacing: float | None = Field(default=None, alias="letterSpacing")


class OPFFontScheme(_OPFBase):
    scheme: str | None = None
    heading: OPFFont | None = None
    body: OPFFont | None = None
    accent: OPFFont | None = None
    code: OPFFont | None = None


class OPFDimensions(_OPFBase):
    preset: str | None = None
    width_inches: float | None = Field(default=None, alias="widthInches")
    height_inches: float | None = Field(default=None, alias="heightInches")


class OPFGradientStop(_OPFBase):
    color: str
    position: float


class OPFGradient(_OPFBase):
    angle: float
    stops: list[OPFGradientStop]


class OPFBackgroundImage(_OPFBase):
    url: str
    fit: Literal["cover", "contain", "tile"]


class OPFBackground(_OPFBase):
    type: Literal["solid", "gradient", "image", "pattern"]
    color: str | None = None
    gradient: OPFGradient | None = None
    image: OPFBackgroundImage | None = None
    opacity: float | None = None


class OPFPosition(_OPFBase):
    x: float | None = None
    y: float | None = None
    anchor: str | None = None


class OPFBrandLogo(_OPFBase):
    url: str
    position: OPFPosition | None = None
    width_inches: float | None = Field(default=None, alias="widthInches")


class OPFBrandWatermark(_OPFBase):
    url: str
    opacity: float | None = None


class OPFBrand(_OPFBase):
    logo: OPFBrandLogo | None = None
    watermark: OPFBrandWatermark | None = None
    company_name: str | None = Field(default=None, alias="companyName")


class OPFLayoutPreferences(_OPFBase):
    density: str | None = None
    alignment: str | None = None
    animations: bool | None = None
    slide_numbers: bool | None = Field(default=None, alias="slideNumbers")


class OPFDesign(_OPFBase):
    theme: str | None = None
    colors: OPFColorScheme | None = None
    fonts: OPFFontScheme | None = None
    dimensions: OPFDimensions | None = None
    background: OPFBackground | None = None
    brand: OPFBrand | None = None
    layout_preferences: OPFLayoutPreferences | None = Field(
        default=None, alias="layoutPreferences"
    )


# ─── Elements ────────────────────────────────────────────────────────


class OPFSize(_OPFBase):
    width: float | None = None
    height: float | None = None
    unit: Literal["in", "%", "px"] | None = None


class OPFAnimation(_OPFBase):
    type: Literal["appear", "fadeIn", "slideIn", "zoomIn", "typewriter", "custom"]
    delay: float | None = None
    duration: float | None = None
    order: int | None = None


class OPFTextRun(_OPFBase):
    text: str
    bold: bool | None = None
    italic: bool | None = None
    underline: bool | None = None
    strikethrough: bool | None = None
    color: str | None = None
    font_size: float | None = Field(default=None, alias="fontSize")
    font_family: str | None = Field(default=None, alias="fontFamily")
    link: str | None = None
    superscript: bool | None = None
    subscript: bool | None = None


class OPFTextContent(_OPFBase):
    text: str | None = None
    runs: list[OPFTextRun] | None = None
    markdown: str | None = None
    bullets: list[Union[str, "OPFTextContent"]] | None = None
    numbered: list[Union[str, "OPFTextContent"]] | None = None


class OPFTextStyle(_OPFBase):
    font_size: float | None = Field(default=None, alias="fontSize")
    font_family: str | None = Field(default=None, alias="fontFamily")
    color: str | None = None
    alignment: Literal["left", "center", "right", "justify"] | None = None
    line_height: float | None = Field(default=None, alias="lineHeight")
    letter_spacing: float | None = Field(default=None, alias="letterSpacing")
    text_transform: (
        Literal["none", "uppercase", "lowercase", "capitalize"] | None
    ) = Field(default=None, alias="textTransform")
    vertical_alignment: Literal["top", "middle", "bottom"] | None = Field(
        default=None, alias="verticalAlignment"
    )


class _OPFElementBase(_OPFBase):
    id: str
    position: OPFPosition | None = None
    size: OPFSize | None = None
    animation: OPFAnimation | None = None
    slot: str | None = None


class OPFTextElement(_OPFElementBase):
    type: Literal["text"] = "text"
    content: OPFTextContent
    style: OPFTextStyle | None = None


class OPFImageElement(_OPFElementBase):
    type: Literal["image"] = "image"
    src: str
    alt: str | None = None
    fit: Literal["cover", "contain", "fill", "none"] | None = None
    border_radius: float | None = Field(default=None, alias="borderRadius")
    shadow: bool | None = None
    caption: str | None = None


class OPFShapeStroke(_OPFBase):
    color: str
    width: float


class OPFShapeElement(_OPFElementBase):
    type: Literal["shape"] = "shape"
    shape: Literal[
        "rectangle",
        "circle",
        "ellipse",
        "triangle",
        "arrow",
        "line",
        "star",
        "callout",
        "custom",
    ]
    fill: str | None = None
    stroke: OPFShapeStroke | None = None
    corner_radius: float | None = Field(default=None, alias="cornerRadius")
    text: OPFTextContent | None = None
    path: str | None = None


class OPFChartDataset(_OPFBase):
    label: str
    values: list[float]
    color: str | None = None
    type: str | None = None


class OPFChartData(_OPFBase):
    labels: list[str] | None = None
    datasets: list[OPFChartDataset]


class OPFChartAxis(_OPFBase):
    label: str | None = None
    min: float | None = None
    max: float | None = None
    format: str | None = None


class OPFChartLegend(_OPFBase):
    position: Literal["top", "bottom", "left", "right"]


class OPFChartOptions(_OPFBase):
    title: str | None = None
    legend: Union[bool, OPFChartLegend] | None = None
    show_values: bool | None = Field(default=None, alias="showValues")
    show_grid: bool | None = Field(default=None, alias="showGrid")
    stacked: bool | None = None
    y_axis: OPFChartAxis | None = Field(default=None, alias="yAxis")
    x_axis: OPFChartAxis | None = Field(default=None, alias="xAxis")


class OPFChartElement(_OPFElementBase):
    type: Literal["chart"] = "chart"
    chart_type: Literal[
        "bar",
        "column",
        "line",
        "pie",
        "donut",
        "area",
        "scatter",
        "radar",
        "waterfall",
        "funnel",
        "treemap",
        "combo",
    ] = Field(alias="chartType")
    data: OPFChartData
    options: OPFChartOptions | None = None


class OPFTableStyle(_OPFBase):
    header_background: str | None = Field(default=None, alias="headerBackground")
    header_color: str | None = Field(default=None, alias="headerColor")
    striped_rows: bool | None = Field(default=None, alias="stripedRows")
    border_color: str | None = Field(default=None, alias="borderColor")
    compact: bool | None = None


class OPFTableElement(_OPFElementBase):
    type: Literal["table"] = "table"
    headers: list[str] | None = None
    rows: list[list[Any]]
    style: OPFTableStyle | None = None


class OPFVideoElement(_OPFElementBase):
    type: Literal["video"] = "video"
    src: str
    poster: str | None = None
    autoplay: bool | None = None


class OPFCodeElement(_OPFElementBase):
    type: Literal["code"] = "code"
    code: str
    language: str | None = None
    theme: Literal["dark", "light"] | None = None
    show_line_numbers: bool | None = Field(default=None, alias="showLineNumbers")


class OPFPlaceholderElement(_OPFElementBase):
    type: Literal["placeholder"] = "placeholder"
    prompt: str
    expected_type: Literal["text", "image", "chart", "table"] | None = Field(
        default=None, alias="expectedType"
    )


class OPFGroupElement(_OPFElementBase):
    type: Literal["group"] = "group"
    children: list["OPFElement"]


OPFElement = Annotated[
    Union[
        OPFTextElement,
        OPFImageElement,
        OPFShapeElement,
        OPFChartElement,
        OPFTableElement,
        OPFVideoElement,
        OPFCodeElement,
        OPFGroupElement,
        OPFPlaceholderElement,
    ],
    Field(discriminator="type"),
]


# ─── Slide & Document ────────────────────────────────────────────────


class OPFTransition(_OPFBase):
    type: Literal["none", "fade", "slide", "push", "wipe", "morph", "zoom"]
    duration: float | None = None
    direction: Literal["left", "right", "up", "down"] | None = None


class OPFSlide(_OPFBase):
    id: str
    layout: str
    design_overrides: OPFDesign | None = Field(default=None, alias="designOverrides")
    elements: list[OPFElement]
    notes: str | None = None
    transition: OPFTransition | None = None
    hidden: bool | None = None
    section: str | None = None
    beat: str | list[str] | None = None
    """Optional reference to one or more narrative beats (each value matches
    an ``id`` from ``meta.narrative.beats`` or the resolved template). A
    single string declares the slide's primary beat; a list declares that
    one slide covers multiple beats — useful when shorter decks fold beats
    together (e.g. ``["problem", "why-now"]``). Slides without a beat are
    valid; multiple slides may share a beat."""


class OPFDocument(_OPFBase):
    schema_url: Literal["https://pptx.dev/schema/opf/v1"] = Field(
        default="https://pptx.dev/schema/opf/v1",
        alias="$schema",
    )
    meta: OPFMeta
    design: OPFDesign
    slides: list[OPFSlide]
    extensions: dict[str, Any] | None = None


OPFTextContent.model_rebuild()
OPFGroupElement.model_rebuild()
OPFSlide.model_rebuild()
OPFDocument.model_rebuild()


__all__ = [
    "OPFAnimation",
    "OPFBackground",
    "OPFBackgroundImage",
    "OPFBrand",
    "OPFBrandLogo",
    "OPFBrandWatermark",
    "OPFChartAxis",
    "OPFChartData",
    "OPFChartDataset",
    "OPFChartElement",
    "OPFChartLegend",
    "OPFChartOptions",
    "OPFCodeElement",
    "OPFColorScheme",
    "OPFDesign",
    "OPFDimensions",
    "OPFDocument",
    "OPFElement",
    "OPFFont",
    "OPFFontScheme",
    "OPFGradient",
    "OPFGradientStop",
    "OPFGroupElement",
    "OPFImageElement",
    "OPFLayoutPreferences",
    "OPFMeta",
    "OPFNarrative",
    "OPFNarrativeBeat",
    "OPFPlaceholderElement",
    "OPFPosition",
    "OPFShapeElement",
    "OPFShapeStroke",
    "OPFSize",
    "OPFSlide",
    "OPFTableElement",
    "OPFTableStyle",
    "OPFTextContent",
    "OPFTextElement",
    "OPFTextRun",
    "OPFTextStyle",
    "OPFTransition",
    "OPFVideoElement",
]
